src/Neo/SmartContract/Native/NativeContract.cs

```diff
@@ -235,7 +296,15 @@ internal async void Invoke(ApplicationEngine engine, byte version)
                 if (version != 0)
                     throw new InvalidOperationException($"The native contract of version {version} is not active.");
                 ExecutionContext context = engine.CurrentContext;
-                ContractMethodMetadata method = methods[context.InstructionPointer];
+                if (currentAllowedMethods.Count == 0)
+                {
+                    // First call we need to cache it
+                    uint index = engine.PersistingBlock is not null ? engine.PersistingBlock.Index : Ledger.CurrentIndex(engine.Snapshot);
+                    _ = GetContractState(engine.ProtocolSettings, index);
+                }
+                ContractMethodMetadata method = currentAllowedMethods[context.InstructionPointer];
```

Comment:
I want to remove `currentAllowedMethods` because is not snapshot dependent

Reply:
It's easy to do it if we change the offset and the native contract's nefScript, it's used internally only, but I think that @roman-khimov and @Liaojinghui won't be on favor of this change 😅

Reply:
I think this should be handled via native contract updates at the specified HF height with the new contract NEF/manifest having new contract methods. This would be pretty natural (old contract doesn't have it, new one does), simple and will allow to preserve state root compatibility.